### PR TITLE
Notifications API

### PR DESCRIPTION
**Pull request by Tunous**
Required for https://github.com/slapperwan/gh4a/issues/175.

- [x] Notification reason constants
- [x] Test marking all notifications as read
- [x] Test marking repository notifications as read
- [x] Test loading thread subscription (Probably won't be used)
- [x] Test marking thread subscription as (not)subscribed
- [x] Test marking thread subscription as (not)ignored
- [x] Add listing of notifications in a single repository (Probably won't be used in my initial implementation but might be useful in future)
- [x] Add deleting thread subscription (No idea what that is for)

### CODE REVIEW

File: org.eclipse.egit.github.core/src/org/eclipse/egit/github/core/client/GitHubClient.java [hunk 1]
```diff
@@ -12,17 +12,7 @@
...
-import static java.net.HttpURLConnection.HTTP_CONFLICT;
-import static java.net.HttpURLConnection.HTTP_CREATED;
-import static java.net.HttpURLConnection.HTTP_FORBIDDEN;
-import static java.net.HttpURLConnection.HTTP_GONE;
-import static java.net.HttpURLConnection.HTTP_INTERNAL_ERROR;
-import static java.net.HttpURLConnection.HTTP_NOT_FOUND;
-import static java.net.HttpURLConnection.HTTP_NO_CONTENT;
-import static java.net.HttpURLConnection.HTTP_OK;
-import static java.net.HttpURLConnection.HTTP_UNAUTHORIZED;
+import static java.net.HttpURLConnection.*;
```
Comment (maniac103) [inline org.eclipse.egit.github.core/src/org/eclipse/egit/github/core/client/GitHubClient.java]: AS probably did that automatically, but can you please revert that part and import only the new code?

Comment (Tunous) [inline org.eclipse.egit.github.core/src/org/eclipse/egit/github/core/client/GitHubClient.java]: Yep, I'll fix that.

File: org.eclipse.egit.github.core/src/org/eclipse/egit/github/core/service/NotificationService.java [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+	}
+
+	/**
+	 * Mark all notifications in a repository as read
+	 *
+	 * @param repoId the id of the repository
+	 * @throws IOException if something went wrong
+	 */
+	public void markNotificationsAsRead(String repoId)
+			throws IOException {
```
Comment (maniac103) [inline org.eclipse.egit.github.core/src/org/eclipse/egit/github/core/service/NotificationService.java]: Could probably avoid the line wrap here and in similar short lines.

File: org.eclipse.egit.github.core/src/org/eclipse/egit/github/core/Notification.java [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+import java.util.Date;
+
+public class Notification implements Serializable {
+
+	/**
+	 * serialVersionUID
+	 */
+	private static final long serialVersionUID = 989467705847081308L;
+
+	private String id;
```
Comment (maniac103) [inline org.eclipse.egit.github.core/src/org/eclipse/egit/github/core/Notification.java]: Shouldn't this be a long?

File: org.eclipse.egit.github.core/src/org/eclipse/egit/github/core/service/NotificationService.java [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+		return createPageIterator(request);
+	}
+
+	/**
+	 * Mark a thread as read
+	 *
+	 * @param threadId the id of the thread subscription
+	 * @throws IOException if something went wrong
+	 */
+	public void markThreadAsRead(String threadId)
```
Comment (maniac103) [inline org.eclipse.egit.github.core/src/org/eclipse/egit/github/core/service/NotificationService.java]: Should threadId be a long here and below?

**Comment by Tunous** [comment]
Loading thread subscriptions works but looks like the GitHub API returns it only for few reasons and I don't see an exact list of which ones (just few examples). Not sure if it's a good idea to leave that method implemented... It won't be used by my implementation.

File: org.eclipse.egit.github.core/src/org/eclipse/egit/github/core/Notification.java (hunk 3 above)
Comment (Tunous) [inline org.eclipse.egit.github.core/src/org/eclipse/egit/github/core/Notification.java]: Probably not, [API](https://developer.github.com/v3/activity/notifications/#list-your-notifications) displays it in double quotes. Or is it somehow converted on egit side?

**Comment by maniac103** [comment]
> Loading thread subscriptions works but looks like the GitHub API returns it only for few reasons and I don't see an exact list of which ones (just few examples).

https://developer.github.com/v3/activity/notifications/#notification-reasons ?

File: org.eclipse.egit.github.core/src/org/eclipse/egit/github/core/Notification.java (hunk 3 above)
Comment (maniac103) [inline org.eclipse.egit.github.core/src/org/eclipse/egit/github/core/Notification.java]: No, you're right. Github deserves a *facepalm* for this though.

**Comment by Tunous** [comment]
> https://developer.github.com/v3/activity/notifications/#notification-reasons

These are all the possible reasons but not all of them make the get subscription work. For example "subscribed" doesn't work but "manual" does.

The only information is this:
> Note that subscriptions are only generated if a user is participating in a conversation--for example, they've replied to the thread, were @mentioned, or manually subscribe to a thread.

But that's not precise.

**Comment by maniac103** [comment]
Maybe I am missing something, but what exactly do you mean by 'make the get subscription work'?

**Comment by Tunous** [comment]
Return some information instead of not found error.

**Comment by Tunous** [comment]
Talking about this one: https://developer.github.com/v3/activity/notifications/#get-a-thread-subscription

**Comment by maniac103** [comment]
But why would we care what exactly triggers subscription creation on the server side when querying for subscription presence?

**Comment by Tunous** [comment]
> But why would we care what exactly triggers subscription creation on the server side when querying for subscription presence?

Damn, that's a good question. I think I got a bit lost trying to figure out how this one is supposed to return a valid response and didn't think about that. That means that it should be all good :)

**Comment by Tunous** [comment]
Should be ready now.

**Pull request merged**